sdk/core/core-client/src/authorizeRequestOnClaimChallenge.ts

```diff
@@ -0,0 +1,96 @@
+import { decodeString, uint8ArrayToString } from "./base64";
+
+/**
+ * Converts: `Bearer a="b", c="d", Bearer d="e", f="g"`.
+ * Into: `[ { a: 'b', c: 'd' }, { d: 'e', f: 'g' } ]`.
+ *
+ * @internal
+ */
+export function parseCAEChallenge(challenges: string): any[] {
+  return `, ${challenges.trim()}`
+    .split(", Bearer ")
+    .filter((x) => x)
+    .map((challenge) =>
+      `${challenge.trim()}, `
+        .split('", ')
+        .filter((x) => x)
+        .map((keyValue) => (([key, value]) => ({ [key]: value }))(keyValue.trim().split('="')))
+        .reduce((a, b) => ({ ...a, ...b }), {})
+    );
+}
+
+/**
+ * CAE Challenge structure
+ */
+export interface CAEChallenge {
+  scope: string;
+  claims: string;
+}
+
+/**
+ * This function can be used as a callback for the `bearerTokenAuthenticationPolicy` of `@azure/core-rest-pipeline`, to support CAE challenges:
+ * [Continuous Access Evaluation](https://docs.microsoft.com/azure/active-directory/conditional-access/concept-continuous-access-evaluation).
+ *
+ * Call the `bearerTokenAuthenticationPolicy` with the following options:
+ *
+ * ```ts
+ * import { bearerTokenAuthenticationPolicy } from "@azure/core-rest-pipeline";
+ * import { authorizeRequestOnClaimChallenge } from "@azure/core-client";
+ *
+ * const bearerTokenAuthenticationPolicy = bearerTokenAuthenticationPolicy({
+ *   authorizeRequestOnChallenge: authorizeRequestOnClaimChallenge
+ * });
+ * ```
+ *
+ * Once provided, the `bearerTokenAuthenticationPolicy` policy will internally handle Continuous Access Evaluation (CAE) challenges.
+ * When it can't complete a challenge it will return the 401 (unauthorized) response from ARM.
+ *
+ * Example challenge with claims:
+ *
+ * ```
+ * Bearer authorization_uri="https://login.windows-ppe.net/", error="invalid_token",
+ * error_description="User session has been revoked",
+ * claims="eyJhY2Nlc3NfdG9rZW4iOnsibmJmIjp7ImVzc2VudGlhbCI6dHJ1ZSwgInZhbHVlIjoiMTYwMzc0MjgwMCJ9fX0="
+ * ```
+ */
+export async function authorizeRequestOnClaimChallenge(
+  onChallengeOptions: AuthorizeRequestOnChallengeOptions
+): Promise<boolean> {
+  const { scopes, response } = onChallengeOptions;
+
+  const challenge = response.headers.get("WWW-Authenticate");
+  if (!challenge) {
+    console.log(
+      `The WWW-Authenticate header was missing. Failed to perform the Continuous Access Evaluation authentication flow.`
+    );
+    return false;
+  }
+  const challenges: CAEChallenge[] = parseCAEChallenge(challenge) || [];
+
+  const parsedChallenge = challenges.find((x) => x.claims);
+  if (!parsedChallenge) {
+    console.log(
+      `The WWW-Authenticate header was missing the necessary "claims" to perform the Continuous Access Evaluation authentication flow.`
+    );
+    return false;
+  }
+
+  const accessToken = await onChallengeOptions.getAccessToken(
+    parsedChallenge.scope ? [parsedChallenge.scope] : scopes,
+    {
+      claims: uint8ArrayToString(decodeString(parsedChallenge.claims))
```

Comment:
I'm confused... `decodeString` changes the string into a Uint8Array... but then `uint8ArrayToString` converts it back to string?

Reply:
this is really base64 to string. Do we have a function that does it directly?

Reply:
Isn't base64 -> string what `atob` does?

Reply:
I’m not sure why I had the idea that `atob` was unreliable. I’ll try it!

Reply:
wait, atob doesn’t exist in Node? 🤔 typescript says its undefined.

Reply:
core-client already has some base64 related functions that I’m using. In this PR I’m only adding one more. In Node it looks like this:

```ts
export function uint8ArrayToString(ab: Uint8Array): string {
  return ab.toString();
}
```

in the browser is a bit more expensive though:

```ts

export function uint8ArrayToString(ab: Uint8Array): string {
  return decodeURIComponent(Array.from(ab, (item) => String.fromCharCode(item)).join(""));
}
```

(notice that I removed the `escape` thing, that was a mistake)

Reply:
Okay I think I understand my confusion here.

Let's consider two scenarios that use base64:

1. Encoding a simple string, maybe with some non-letter characters
2. Encoding an arbitrary byte array

We can view 2 as a more general case of 1 as you can think of a utf8 string as a uint8array. 

What is interesting to me is we have two encode functions (`encodeString` and `encodeByteArray`) but we only have a single `decodeString` which always returns the byte array, which necessitates you adding this function.

But really here we want decodeString to return a string, not an array, and we're actually doing a lot of unnecessary work in the browser version when calling `uint8ArrayToString(decodeString(value))`:

```ts

// let's look at the function bodies
export function decodeString(value: string): Uint8Array {
  // byteString is actually the value we want in this case! But we don't return it, instead we encode it as an array.
  const byteString = atob(value);
  const arr = new Uint8Array(byteString.length);
  for (let i = 0; i < byteString.length; i++) {
    arr[i] = byteString.charCodeAt(i);
  }
  return arr;
}

// now because we encoded it as an array... we have to decode it again
export function uint8ArrayToString(ab: Uint8Array): string {
  return decodeURIComponent(Array.from(ab, (item) => String.fromCharCode(item)).join(""));
}
```

Whereas what I really want is a function like

```
export function decodeStringToString(value: string): string {
  return atob(value);
}
```

So can we add *that* method instead? The node version is also trivial: 

```
export function decodeStringToString(value: string): string {
  return Buffer.from(value, "base64").toString();
}
```

Reply:
Ok sounds good! can do :) thank you for the thorough rundown 🙏

Reply:
I have pushed the update. My only criticism of this approach is that it isn’t reusing a function that seems more fundamental, but it seems silly to argue that in retrospective.